# What's a Program?

A set of instructions you give to a computer to do a specific task.

Python is very similar to English. Easy to learn, Very powerful language. This is mainly used in AI/ML, build apps, write games, ethically hack.

Syntax: Rules by which you are supposed to arrange words/pieces for it all to make sense.

"A boy kicked the ball". "Ball kicked the boy." Syntax is very important.

### print statement

This is like you telling the computer to say what you want it to say

In [3]:
print("Hi, this is a porgramming class")

Hi, this is a porgramming class


In [4]:
print(20)

20


In [5]:
print(True)

True


# Data Types

A group of values which have a specific set of actions associated with them. A generic umbrella for specific kinds of values.

4 principal data types: Integer, String, Boolean, Float (decimal values)

Integer: There are basically from -infiinty to infinity (in most cases in programing langues is -2*31 to (2^31)-1.

Float: These are numbers just like integers, but they have the decimal part to them as well.

String: These are set of charcaters put together. "a", "hello", "iaufwye123"

Boolean: These are values that are either True or False. They are answers to conditions/conditional statements.

In [6]:
# These are all integers
a = 20
b = 30
c = -100
# These are floats
f1 = 2.3
f2 = 6.0
f3 = -19.5

In [7]:
# Booleans only have 2 values, really
b1 = True
b2 = False
# These are strings
first = "sonething"
# Single or double quotes are essential to identifying strings
second = "7812387416qweu iofhqweiouf*!&%@#*^$"

In [8]:
# Should return 50
a+b

50

In [9]:
a-b # This should give you -10

-10

In [10]:
a*b # This should give you 600

600

In [11]:
a/b # This should give you liek 0.66666667

0.6666666666666666

In [13]:
# 2 is the base, 3 is the exponent
2**3 # THis should give you 8, as "**" is used for exponential calculations

8

### Variables and Constants

Variables are the ones whose value can change at any time. Constants do not change values during the execution of a program

# Rules to make your variables

1. You cannot start it with a number.
2. You can have any characters from a-z, A-Z, to start them.
3. You can have numbers or special chracters like "_" in them. You cannot have spaces in variables. 
4. var_1Real --> Valid
5. 1i8 --> invalid variable name

Input --> What does into the program. You send some data in to get some data out. This is the data that is sent in.

Output --> What comes out of the program. This is the data that comes out.

In [15]:
a = 20
b = 20
# a, b is the input
# a*b, the 400 that comes out, is the output
print(a*b)

400


### Condittional statements --> These statements allow you to execute or perform certain tasks based on certain situations/conditions

In [16]:
# if, else, elif (else if)
# if gives you the first condition --> if the condition is True, we will go ahead with the execution. If it is not, we will not execute the statements

In [21]:
# You normally tab blocks of code to group them by conditions or by loops.
if a > b:
    print("A is Bigger than B")
elif a == b: # This checks if the values are equal
    print("They are equal")
else:
    print("Okay idk whats up")

They are equal


# Loops

They allow you to perform certain tasks over and over again. When you want to repeated do tasks. For loop, and a while loop, I will get into these in a future class.

In [23]:
for ind in range(1, 11):
    print(ind)

1
2
3
4
5
6
7
8
9
10


# Arthemetic operators:
+, -, *, /, **

# Logical operators"
">", "<", "==", ">=", "<=", and, or

In [22]:
# Check if a give number is bigger than the other. You can make up your own numbers. If your first numbers is bigger than the second, print out that First Wins.
# Otherwise, print out that Second wins

Comments: They are for humans, and are not included in the execution of the code. They increase the readablity of the code.

# For Loop
### This is when you know how many times you want to run something

### This loop runs from start to stop - 1
for var in range(start, stop, step) - step is the increment; you skip by a given number every time

In [2]:
for ind in range(0, 20, 2):
    print(ind)

0
2
4
6
8
10
12
14
16
18


# Print all odd numbers from 1 to 30

In [7]:
for ind in range(1, 30, 2):
    print(ind)

1
3
5
7
9
11
13
15
17
19
21
23
25
27
29


In [10]:
# don't always have to use ind, use any valid variable name
for ind in range(5, 105, 25):
    print(ind)

5
30
55
80


In [15]:
# Print the multiples of 7 between 2 and 80
for ind in range(7, 80, 7):
    print(ind)

7
14
21
28
35
42
49
56
63
70
77


# While loop
### You run this loop when you don't know how many time you want to do something

While (condition):
    
    ### Do something

In [17]:
### While loops can result in infinite loops if the conditions are not altered.

In [18]:
# While loop breaks when the condition is false
number = 20
while number < 0:
    print(number)
    number -= 1

In [19]:
for ind in range(10, -1, -1):
    print(ind)

10
9
8
7
6
5
4
3
2
1
0


In [21]:
# Default step is considered 1
for ind in range(1, 11):
    print(ind)

1
2
3
4
5
6
7
8
9
10


In [27]:
# Multiples of 4 between 400 and 200 in reverse order
# Limit here is 203, and the last number printed was 204. It is still less than the limit, so we go one step further
for ind in range(400, 199, -4):
    print(ind)

400
396
392
388
384
380
376
372
368
364
360
356
352
348
344
340
336
332
328
324
320
316
312
308
304
300
296
292
288
284
280
276
272
268
264
260
256
252
248
244
240
236
232
228
224
220
216
212
208
204
200


In [29]:
# This is the modulo operator, which gives you the remainder of a given number divided by another
for ind in range(200, 401):
    if ind % 4 == 0:
        print(ind)

200
204
208
212
216
220
224
228
232
236
240
244
248
252
256
260
264
268
272
276
280
284
288
292
296
300
304
308
312
316
320
324
328
332
336
340
344
348
352
356
360
364
368
372
376
380
384
388
392
396
400


In [30]:
print(16 % 5)

1


In [32]:
# Write a program that would tell me if a number is even or odd
number = 23
if number % 2 == 0:
    print("Even")
else:
    print("Odd")

Odd


# Data

Information that a computer uses to do some tasks or it is what it gives out to you. Input and output is a form of data in computer programs.

A program takes in a particular input and gives you a specific output.

### Data Structures are things that hold data. Today we will learn about Lists and Dictionaries

Lists: They look like ```[]``` and allow you to hold things inside it. Lists are accessed by indices. They are 0 indexed - the first index is 0, the second is 1, and you can go on.

In [8]:
list1 = ["Apple", 1, 2.0, False, 9, 55]

In [4]:
list1[2]

2.0

### If you use the ```len()``` function, you can tell how many elements are in the list

In [9]:
len(list1)

6

In [10]:
print(list1)

['Apple', 1, 2.0, False, 9, 55]


### We can use the for loop to iterate through elements in the list

In [11]:
for element in list1:
    print(element)

Apple
1
2.0
False
9
55


In [13]:
for ind in range(len(list1)):
    print(list1[ind])

Apple
1
2.0
False
9
55


In [14]:
## Exercise problem

### Define a list with 5-7 elements. I want you to print out how many elements are in the list, print out the list, and print out each element in the list

In [15]:
list2 = [1, 2, 3, 4, 5, 6]

In [16]:
print(len(list2))

6


In [18]:
print(list2)

[1, 2, 3, 4, 5, 6]


In [19]:
for elem in list2:
    print(elem)

1
2
3
4
5
6


In [20]:
for ind in range(len(list2)):
    print(list2[ind])

1
2
3
4
5
6


### append() --> Adds elements to the end of a list

### pop() --> Removes elements from the end of a list

In [23]:
empty_list = []
print("List before adding anything: ", empty_list)
empty_list.append(20)
empty_list.append(40)
print(empty_list)
empty_list.pop()
print(empty_list)

List before adding anything:  []
[20, 40]
[20]


### Print out numbers from 1 to 10. I want you to add all these numbers to a list and print the list

In [31]:
# This is the list I am gonna add stuff to
result_list = []
# This allows me to access numbers from 1 to 10
for ind in range(1, 11):
    result_list.append(ind)
    print(result_list)
print(result_list)

[1]
[1, 2]
[1, 2, 3]
[1, 2, 3, 4]
[1, 2, 3, 4, 5]
[1, 2, 3, 4, 5, 6]
[1, 2, 3, 4, 5, 6, 7]
[1, 2, 3, 4, 5, 6, 7, 8]
[1, 2, 3, 4, 5, 6, 7, 8, 9]
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


### Pop 3 elements from result_list

In [32]:
print(result_list)
for ind in range(3):
    result_list.pop()
    print(result_list)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
[1, 2, 3, 4, 5, 6, 7, 8, 9]
[1, 2, 3, 4, 5, 6, 7, 8]
[1, 2, 3, 4, 5, 6, 7]


### Print multiples of 13 between 2 and 140

In [2]:
for ind in range(2, 140):
    if ind % 13 == 0:
        print(ind)

13
26
39
52
65
78
91
104
117
130


In [3]:
for ind in range(13, 140, 13):
    print(ind)

13
26
39
52
65
78
91
104
117
130


# Reverse the list on Paper

In [4]:
sample = [2, 3, 4, 1, 5]

In [6]:
result = []
for ind in range(len(sample)-1, -1, -1):
    print(ind)
    result.append(sample[ind])
result

4
3
2
1
0


[5, 1, 4, 3, 2]

# Swapping values between 2 variables

In [7]:
a = 10
b = 5

In [8]:
b = a
a = b
a, b

(10, 10)

In [15]:
a = 10
b = 5
c = b

In [16]:
c

5

In [17]:
b = a
b

10

In [18]:
a = c
a, b

(5, 10)

In [19]:
a, b = b, a
a, b

(10, 5)

In [20]:
result

[5, 1, 4, 3, 2]

In [21]:
left = 0
right = len(result)-1
while left < right:
    result[left], result[right] = result[right], result[left]
    left += 1
    right -= 1
result

[2, 3, 4, 1, 5]

# Reverse the list: [1, 2, 3, 4, 5, 6]

In [22]:
result = [1, 2, 3, 4, 5, 6]
left = 0
right = len(result)-1
while left < right:
    result[left], result[right] = result[right], result[left]
    left += 1
    right -= 1
result

[6, 5, 4, 3, 2, 1]

Write code to tell if a number is even or not

# Make you own list of movies/TV shows. Pick the middle of the list every time, and add it to the end. For ever 3 times you add to the end, Remove the last 2 elements from the list. Do these things 5 times.

In [ ]:
# 1 iteration. Do this 5 times.
[a, b, c]
[a, b, c, b]
[a, b, c, b, b]
[a, b, c, b, b, b]
[a, b, c, b]

In [15]:
movies = ["Movie 1", "Movie 2", "Movie 3"]

In [11]:
middle = len(movies)//2
middle_movie = movies[middle]
middle_movie

'Movie 3'

In [12]:
movies.append(middle_movie)
movies

['Movie 1', 'Movie 2', 'Movie 3', 'Movie 2', 'Movie 3', 'Movie 3']

In [13]:
movies.pop()
movies.pop()

'Movie 3'

In [14]:
movies

['Movie 1', 'Movie 2', 'Movie 3', 'Movie 2']

In [17]:
movies = ["Movie 1", "Movie 2", "Movie 3"]
counter = 1
for ind in range(5):
    if counter % 4 == 0:
        movies.pop()
        movies.pop()
        counter = 0
    else:
        middle = len(movies)//2
        middle_movie = movies[middle]
        movies.append(middle_movie)
        print(movies)
        counter += 1

['Movie 1', 'Movie 2', 'Movie 3', 'Movie 2']
['Movie 1', 'Movie 2', 'Movie 3', 'Movie 2', 'Movie 3']
['Movie 1', 'Movie 2', 'Movie 3', 'Movie 2', 'Movie 3', 'Movie 3']


### Write a program to find the volume of rectangular prism, which was designed by melting a cube with side 5 cm.

In [18]:
def volume(side):
    return side * side * side

In [19]:
volume(5)

125

### You have a list of depths of hippos (how deep into the water they are. Based on each depth, you must print out if the hippo is low, medium, or deep.
### If depth is less than 50 cm, its low. It it is between 50 and 100 cm, its medium. Or else its high.

In [20]:
def depth_status(depths):
    # depths = [10, 20, 50, 90, 100, 40, 65, 75]
    #
    # you can print out your answers
    # GO over the list
    for depth in depths:
        if depth < 50:
            print("Low")
        elif depth >= 50 and depth <= 100:
            print("Medium")
        else:
            print("High")

In [21]:
depth_status([10, 20, 50, 90, 100, 40, 65, 75])

Low
Low
Medium
Medium
Medium
Low
Medium
Medium


# You will get a number. I want you to find the sum of its digits, until the final result is less than 10 (single digit).

In [14]:
def sum_of_digs(number):
    # The final answer should be in a single digit.
    # Example: 192
    # Iteration 1: 1 + 9 + 2 = 12
    # Iteration 2: 1 + 2 = 3
    # We return 3
    res = 0
    while number > 0 or res > 9:
    #while number > 0:
        if number == 0:
            number = res
            res = 0    
        first_digit = number % 10
        res += first_digit
        print(first_digit, res)
        number //= 10
        print(res)
    return res

In [13]:
sum_of_digs(1234)

4 4
4
3 7
7
2 9
9
1 10
10
0 0
0
1 1
1


1

In [21]:
def sum_of_digits_real(number):
    res = number % 9
    if res == 0:
        return 9
    return res

In [22]:
sum_of_digits_real(234)

9

# Given a number, determine if it is prime or not

In [23]:
def check_prime(number):
    # only divisible by 1 and the number
    # use % and if statement
    # Even numbers? 2 is the only even prime number
    # 
    if number < 2:
        return False
    if number % 2 == 0 and number > 2:
        return False
    for num in range(2, number):
        if number % num == 0:
            return False
    return True

In [29]:
check_prime(1)

False

# Given a number, give me its factorial
### 0! = 1, factorial of negative numbers does not exist

In [30]:
def factorial(n):
    # len() the number --> len() cannot be used on numbers
    # If I enter a number less than 0, it does not work
    # 0! = 1
    if n < 0:
        return "Invaild number"
    if n <= 1:
        return 1
    res = 1
    # 2 and other numbers?
    for ind in range(1, n+1):
        res *= ind
    return res

In [35]:
factorial(5)

120

# Ticket out of the Door and Homework

# Mr. Wahoo is trying to climb up n stairs. He can either climb 1 stair or 2 stairs at a time. How many ways can he climb up n stairs?

In [1]:
def wahoo_climb_stairs(n):
    # How many ways can he get to the top?
    # if there is only 1 step, there is one way
    # if there are 2 steps, there are two ways. 1, 1; or 2;
    if n == 1:
        return 1
    if n == 2:
        return 2
    first = 1
    second = 2
    res = 0
    for ind in range(2, n):
        res = first + second
        first = second
        second = res
    return res

In [3]:
wahoo_climb_stairs(4)

5

# You are at (0, 0) in a grid. You can only travel through the edges of the grid, and can only move right or up. How many ways can you get to (m, n), you cannot go backwards on a line.

In [18]:
from math import comb
import math

In [19]:
def ways_in_grid_right_up(m, n):
    return math.comb(m+n, m)

In [20]:
ways_in_grid_right_up(6, 3)

84

# No importing. Implement the combination functions

In [8]:
# Add other functions to help you, as you need
# Return a choose b, given a and b, without importing from math or any other library
# n choose k = n!/((k!)*(n-k)!)
# 2 choose 3 --> Does this work?
def factorial(n):
    res = 1
    for ind in range(1, n+1):
        res *= ind
    return res

def combinations(a, b):
    if a < b or a < 0 or b < 0:
        return "Not Possible"
    return factorial(a)//(factorial(b)*(factorial(a-b)))

In [9]:
combinations(4, 2)

6

# Try to get to (3, 3) in the previous problem

In [14]:
combinations(6, 3)

20

# Implement permutations without importing from Math library

In [1]:
# Some helper
# n!/(n-k)! = permutations, combinations = n!/(k!*(n-k)!)
def factorial(n):
    res = 1
    for ind in range(1, n+1):
        res *= ind
    return res
# Actual thing
def permutations(a, b):
    if a < b or a < 0 or b < 0:
        return "Not Possible"
    return factorial(a)//((factorial(a-b)))

In [2]:
permutations(3, 1)

3

In [7]:
# Vignesh
permutations(4, 4)

24

In [3]:
# Kaysha
permutations(5, 2)

20

In [6]:
# Lucky
permutations(8, 4)

1680

In [5]:
# Pradhyum
permutations(10, 7)

604800

In [4]:
# Abhinav
permutations(8, 3)

336

# Find the sum of a factorial of a number. 5! = 120, digit sum = 1 + 2 + 0 = 3. 10! = 3628800, digit sum = 3 + 2 + 6 + 6 + 8 + 8 + 0 + 0 = 27

In [8]:
def factorial(n):
    res = 1
    for ind in range(1, n+1):
        res *= ind
    return res

def digits_sum(n):
    # hold the result
    res = 0
    # We want to keep going till n is greater than 0
    while n != 0:
        # We extract the last digit of the number
        digit = n % 10
        # Adding the last digit to the sum
        res += digit
        # integer division
        n //= 10
    return res

In [9]:
digits_sum(factorial(10))

27

In [10]:
digits_sum(factorial(7))

9

# We have a list of weights of hippos. We want to get the heaviest and the lightest hippo from the list
### [1000, 2000, 500, 6000, 4000], heaviest = 6000, lightest = 500

# Find if a number is an armstrong number or not.
### Get each digit of the number, raise it to the number of digits, and check if it is the same as the original number.
### 153, 3 digits, so we raise it to the 3rd power. 1^3 + 5^3 + 3^3 = 1 + 125 + 27 = 153. The sum is the same as the original number, so this is an Armstrong number